File: LeetCodePractice/src/DFS/NumberOfEnclaves.java
package DFS;

/*
Given a 2D array A, each cell is 0 (representing sea) or 1 (representing land)

A move consists of walking from one land square 4-directionally to another land square, or off the boundary of the grid.

Return the number of land squares in the grid for which we cannot walk off the boundary of the grid in any number of moves.



Example 1:

Input: [[0,0,0,0],[1,0,1,0],[0,1,1,0],[0,0,0,0]]
Output: 3
Explanation:
There are three 1s that are enclosed by 0s, and one 1 that isn't enclosed because its on the boundary.
Example 2:

Input: [[0,1,1,0],[0,0,1,0],[0,0,1,0],[0,0,0,0]]
Output: 0
Explanation:
All 1s are either on the boundary or can reach the boundary.


Note:

1 <= A.length <= 500
1 <= A[i].length <= 500
0 <= A[i][j] <= 1
All rows have the same size.
 */
public class NumberOfEnclaves {
    int rows, cols;
    int[] dirs = {0, 1, 0, -1, 0};

    public int numEnclaves(int[][] A) {
        rows = A.length;
        cols = A[0].length;
        for (int r = 0; r < rows; r++) {
            dfs(A, r, 0);
            dfs(A, r, cols - 1);
        }
        for (int c = 0; c < cols; c++) {
            dfs(A, 0, c);
            dfs(A, rows - 1, c);
        }
        int res = 0;
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (A[r][c] == 1) {
                    res++;
                }
            }
        }
        return res;
    }

    private void dfs(int[][] A, int r, int c) {
        if (r >= 0 && r < rows && c >= 0 && c < cols && A[r][c] == 1) {
            A[r][c] = 0;
            for (int i = 0; i < dirs.length - 1; i++) {
                dfs(A, r + dirs[i], c + dirs[i + 1]);
            }
        }
    }
}

File: LeetCodePython/NumberOfEnclaves.py
"""
You are given an m x n binary matrix grid, where 0 represents a sea cell and 1 represents a land cell.

A move consists of walking from one land cell to another adjacent (4-directionally) land cell or walking off the boundary of the grid.

Return the number of land cells in grid for which we cannot walk off the boundary of the grid in any number of moves.

Example 1:
Input: grid = [[0,0,0,0],[1,0,1,0],[0,1,1,0],[0,0,0,0]]
Output: 3
Explanation: There are three 1s that are enclosed by 0s, and one 1 that is not enclosed because its on the boundary.

Example 2:
Input: grid = [[0,1,1,0],[0,0,1,0],[0,0,1,0],[0,0,0,0]]
Output: 0
Explanation: All 1s are either on the boundary or can reach the boundary.

Constraints:
m == grid.length
n == grid[i].length
1 <= m, n <= 500
grid[i][j] is either 0 or 1.
"""
from typing import List


class NumberOfEnclaves:
    def numEnclaves(self, grid: List[List[int]]) -> int:
        res = 0
        dirs = [0, 1, 0, -1, 0]
        rows, cols = len(grid), len(grid[0])

        def dfs(r, c):
            grid[r][c] = 0
            for i in range(len(dirs) - 1):
                nb_r, nb_c = r + dirs[i], c + dirs[i + 1]
                if 0 <= nb_r < rows and 0 <= nb_c < cols and grid[nb_r][nb_c] == 1:
                    dfs(nb_r, nb_c)

        for r in range(rows):
            if grid[r][0] == 1:
                dfs(r, 0)
            if grid[r][cols - 1] == 1:
                dfs(r, cols - 1)

        for c in range(cols):
            if grid[0][c] == 1:
                dfs(0, c)
            if grid[rows - 1][c] == 1:
                dfs(rows - 1, c)

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 1:
                    res += 1
        return res